# sciket - learn  복습 

- 데이터 전처리 과정 -> 데이터 셋 분리 -> 모델 생성 및 학습 -> 예측 수행 -> 평가 
- 이해한대로 설명 
    - 필요없는 컬럼 제거 (target값이 같이 붙어있는다던지 등등 제거)
    - 데이터 셋 분리 (data와 target)로 나눈다
    - 모델 생성 : model = regression(예시)
    - 학습 : model.fit(학습할거리 x,y | x_train, y_train , x_test , y_test)
    - 예측 수행 : model.predict(x_test) 보통 x_train,y_train을 입력해 놓은 모델에 x_test의 값을 예측하게 하는편 
    - 평가 : 실제모델 평가 model.fit(y_test, predict된 x_test의 값)

- 처음 배웠던것 LinearRegression 
- 항상 데이터 셋의 쉐잎은 확인해주는것이 좋다 x.shape y.shape  
- 쉐잎이 맞지 않을경우 리쉐잎으로 하나는 이차원 배열을 만들어줘야한다 보통 .reshape(-1,1을 쓴다)
- coef_ 기울기 intercept_ y절편의 값 (coef_가 컬럼 개수만큼 나오는걸로 기억..)
- train_test_split 함수 중요 from sklearn.model_selection 모델에 데이터 전부를 넣어버리면 이 데이터를 입력했을때 자기의 구성성분으로 이루어진 모델에 자기 구성성분을 넣으면 무조건 100%가 나올것이므로 나눠주는것이 필요해 만든 함수 
- x_train, y_train , x_test , y_test (x_valid, y_valid) = train_test_split(x_date,y_data)
기본 형이며 y_data 뒤에 test_size= 0.2 를 붙여준다 
- 이렇게 사용하면 x_data와 y_data의 값을 train 과 test의 형태로 분할해준다 
- test_size 를 사용하면 그 비율을 정할 수 있다 . 0.2 같은경우면 train 0.8 test 0.2 의 비율 
- 쓰지 않은다면 디폴트 값은 0.25로 기억한다. 


- predeict_proba : 예측 확률을 알려준다 
예를들어 결과가 0일 확률을 predict_proba로 뽑고 1일 확률을 predict_proba로 뽑는다면 
둘의 합은 1이 된다 . 

# Simple Linear Regression

- 독립 변수 X와 그에 따라 바뀌는 종속 변수 y로 만든 관계 
- 독립 변수 X_data X_train X_test 등등 (꼭 x여야 하는것이 아님)
- 종속 변수 y_data y_train y_test (X의 타겟값)

# Metric

In [ ]:
# 실습 당뇨병 진행 예측 
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
diabetes = load_diabetes()  # 저장되어있는 데이터 값을 불러옴 
diabetes.data, diabetes.target # data 가 독립변수 target 값이 종속변수 

In [6]:
# 데이터 셋 분리작업 
print(diabetes.data.shape)
print(diabetes.target.shape) # 하나는 백터고 앞에 숫자가 맞으면 서로 train_test_split 가능

(442, 10)
(442,)


In [12]:
# 데이터 셋 분리작업 마무리 
X_train , X_valid , y_train, y_valid = train_test_split(diabetes.data, diabetes.target, test_size=0.2)

In [13]:
# 모델 생성 작업 
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [14]:
# 모델 생성 후 학습 
model.fit(X_train,y_train)

LinearRegression()

In [15]:
y_valid_predict = model.predict(X_valid) # 생성된 모델에 X_valid 테스트 값을 예측값으로 저장 

### 회귀 평가 지표

### MSE (Mean Squared Error)
- 실체값과 예측값의 차이를 제곱해서 평균화 

In [17]:
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_valid_predict, y_valid)
print(mse) # 값이 너무크게 나옴 

3600.5985968812074


### RMSE (Root Mean Squared Error)
- MSE에 루트만 씌운 형태 

In [19]:
# 방법 1
import numpy as np
print(np.sqrt(mse))

# 방법 2 
rmse = mean_squared_error(y_valid_predict, y_valid, squared=False)
print(rmse)

60.004988100000546
60.004988100000546


### MAE (Mean Absolute Error)
- 실제값과 예측값의 차이를 절대값으로 변환해서 평균화 

In [20]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(y_valid_predict, y_valid) 
mae

48.432230104354396

### MAPE (Mean Absolute Percentage Error)

- 실제값에 대한 절대오차 비율의 평균을 퍼센트로 표현 

In [21]:
# 방법1 
def mape(true,pred):
    return np.mean(np.abs((true-pred) / true))
print(mape(y_valid_predict, y_valid))

# 방법2 
from sklearn.metrics import mean_absolute_percentage_error
mean_absolute_percentage_error(y_valid_predict, y_valid)

0.3472310147505409


0.3472310147505409

### SMAPE (Symmetric Mean Absolute Percentage Error)

- 기존 MAPE의 단점을 보완 
- 실제값과 예측값을 절대값으로 변경 후 합으로 나눈다.
- 실제값이 예측값보다 크거나 작을 수도 있다. 


In [22]:
def smape(true,pred):
    error = np.abs(true-pred) / (np.abs(true) + np.abs(pred))
    return np.mean(error)
smape(y_valid_predict, y_valid)

0.17059206280827396

### 결정계수 R2 

- 실제 관측값의 분산대비 예측값의 분산을 계산해 데이터 예측의 정확도 성능을 측청한다.
- 0~1사이의 수로 나타내어지며 1에 가까울수록 100%에 가까워진다.


In [23]:
from sklearn.metrics import r2_score
r2_score(y_valid, y_valid_predict) # r2 스코어는 y_valid값을 앞으로 써줘야한다. 

0.4235282260426284

# 선형 모델 (Linear Models )

### 실습 캘리포니아 주택 가격 데이터 

In [26]:
from sklearn.datasets import fetch_california_housing
import pandas as pd 
california = fetch_california_housing()

In [27]:
california.keys() # 데이터와 타겟 값을 찾기 
california_df = pd.DataFrame(california.data, columns= california.feature_names)
# 켈리포니아 데이터셋 만들어주기 

In [ ]:
import seaborn as sns
sns.pairplot(california_df) # 데이터의 컬름들을 x,y로 삼아 그래프를 그린다 그중 상관있는 그래프 찾기 

In [28]:
model = LinearRegression()
X_train, X_valid, y_train , y_valid = train_test_split(california.data, california.target, 
                                                       test_size = 0.2)

In [30]:
model.fit(X_train, y_train)

LinearRegression()

In [31]:
y_valid_predict = model.predict(X_valid)

In [32]:
print("R2 Score", r2_score(y_valid, y_valid_predict))
print("MSE", mean_squared_error(y_valid_predict, y_valid))

R2 Score 0.6136776329457192
MSE 0.525511110660273


In [35]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(model, X_train, y_train, cv = 4, scoring="neg_root_mean_squared_error")
scores

array([-0.72785379, -0.70896322, -0.73452073, -0.73443761])

#### cross_val_score 
- 교차검증을 하는 함수 모델과 거기에 넣은 xy값을 집어넣고 cv(교차검증 횟수) 돌리면 나온다. 
- 아직 어떤 상황에서 쓰는지 이해가 안감.. 

# 교차 검증 

- 트레이닝셋과 테스트 셋의 비율 : 1만개정도면 80% 트레이닝셋 20% 테스트셋 10만개정도면 99% 트레이닝셋 1% 테스트셋 
- random_state : Random seed 번호, 번호를 지정하면 동일한 데이터 세트로 분리해 준다.
- test_size : 전체 데이터에서 테스트 데이터 세트 크기를 얼마로 샘플링 할것인가(기본 25%)
- train_size : test_size의 남은 값
- shuffle : 데이터를 분리하기 전에 섞을지 결정

## 릿지 회귀

- 내가 기억하는것 : 회귀계수를 0에 가깝게 하지만 0으로 만들지 않음 
- 사용법 : 데이터셋 분리작업까지 똑같이 진행 후 모델만들때 부터 차이남
- 모델 생성 : model = Ridge(alpha = 0.2) * alpha가 규제의 크기 
- 쓰다보면 어느 부분쯤에서 지지부진한 결과가 나오게 되는데 그게 릿지 모델로 개선할 수 있는 한계점이다.


### 캘리포니아 주택 가격 릿지 회귀 간단 실습 

In [ ]:
from sklearn.datasets import fetch_california_housing
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split
california = fetch_california_housing()  # 정보들과 임포트 모듈 불러오고 

# 데이터셋 분리작업 
x_train, x_test, y_train, y_test = train_test_split(california.data, california.target, test_size = 0.2)
# 모델생성 (릿지 규제는 0.2)
model = Ridge(alpha = 0.2)
# 모델 학습 
model.fit(x_train, y_train)
# 그뒤 테스트 
print(f'Train Data Score: {model.score(x_train, y_train)}')
print(f'Test Data Score: {model.score(x_test, y_test)}')

# 라쏘 회귀

- 내가 기억하는것 : 불필요한 회귀계수를 급격하게 감소시켜 0으로 만들고 제거 
- 사용법 : 릿지랑 똑같지만 알파는 0.01 -> 0.001 10의 단위로 바꿀 수 있다.


### 캘리포니아 주택 가격 릿지 회귀 간단 실습 

In [ ]:
from sklearn.linear_model import Lasso
#데이터셋 분리작업 진행 
model = Lasso(alpha = 0.1)
# 모델 학습 
model.fit(x_train, y_train)

# 신축망

- 내가 기억하는것: 릿지, 라쏘 두 모델 둘 다 사용한다.

### 캘리포니아 주택 가격 신축망 실습

In [ ]:
from sklearn.linear_model import ElasticNet

# l1_ratio = 0 L2 규제 = 1 L1 규제 0과 1사이 혼합사용 
model = ElasticNet(alpha=0.001, l1_ratio=0.5)
model.fit(x_train, y_train)

# 학습곡선

- overfiting/underfiting :
- 과대 적합이란 모델이 훈련 세트에서는 좋은 성능을 내지만 검증 세트에서는 낮은 성능을 내
는 경우를 말한다.
- 훈련세트의 정확도가 99% 이고, 검증세트의 정확도가 80% 수준이라면 과대 적합을 의심할
수 있다.
- 과소 적합은 훈련세트와 검증 세트의 성능에는 차이가 크지 않지만 모두 낮은 성능을 내는
경우를 의미한다.

# Logistic Regression

- 이진 데이터 분류 문제입니다.
- 로지스틱 리그레션을 사용하기 위해서는 출력값을 0과 1로 변경해야 함 

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)
# 모델 생성
model = LogisticRegression()
# 모델 학습
model.fit(x_train, y_train)
# 학습 결과
print(f'Train Data Score : {model.score(x_train, y_train)}')
print(f'Test Data Score : {model.score(x_test, y_test)}')
# model의 정확도 평가 
scores = cross_val_score(model, x, y, cv = 10, scoring = 'accuracy') # accur 제대로 분류가 됐는지 측정
print(f'CV mean score: {scores.mean()}') 

### Box plot

- Median : 데이터의 중앙값
- 1사분위(Q1) : Median 을 기준으로 하위 25% 지점
- 3사분위(Q3) : Median 을 기준으로 상위 25% 지점
- IQR (상자길이) : Q3-Q1
- Q1 - 1.5 * IQR or Q3 + 1.5 * IQR 을 whisker라고 함 
- Outliers: Q1 - 1.5 * IQR | Q3 + 1.5 * IQR 밖에 있는데이터 이상치라고도 함 

In [ ]:
# 유방암 데이터 간단 실습 
x, y = load_breast_cancer(return_X_y=True) # xy를 바로 train_test 함수 사용할 수 있게 바꿔줌
x_train, x_test, y_train, y_test = train_test_split(x, y)
# 모델생성
model = LogisticRegression()
# 모델학슴 
model.fit(x_train, y_train)
# 값 출력
print(f'Train Data Score: {model.score(x_train, y_train)}')
print(f'Test Data Score: {model.score(x_test, y_test)}')

# 확률적 경사 하강법

내가 이해한 부분: 기울기가 최소화 되는 방향으로 학습 

# Confusion Matrix
|                     | 실제 True          | 실제 False          |
|---                  |---                 |---                             |
| 분류결과 True       | $$N_{TP}$$         |    $$N_{FP}$$       |
| 분류결과 False      |   $$N_{FN}$$       |   $$N_{TN}$$        | 


#### Precision(정밀도)
- 모델이 True라고 했던것중에 진짜 True의 비율 
- 예측한 결과가 얼마나 실제로 일치하는지 여부 


#### Recall(재현율)
- 실제 True 중에서 모델이 True라고 예측한 것의 비율


- A (9/60 잘된분류 1/30 잘못된분류)

|실제값 |예측값|예측값|
|:---:|:--:|:----:|
|      |암환자|일반환자|
|암환자|9|1|
|일반환자|30|60|

- B (1/70 잘된분류 9/20 잘못된분류 )

|실제값 |예측값|예측값|
|:---:|:--:|:----:|
|      |암환자|일반환자|
|암환자|1|9|
|일반환자|20|70|

#### accuracy(정확도)

- 입력된 데이터에 대해 얼마나 정확하게 예측하는지 나타낸다.
    - 정확도 : 예측결과와 실제값이 동일한 변수 / 전체데이터 개수 = (TP+TN) / (전체)
- 예제로 이해해보기 
    - A병원의 정확도 : 9+60 / (9+60+1+30) = 0.69
    - B병원의 정확도 : 1+70 / 1+9+20+70 = 0.71
    - B병원의 정확도가 미세하게 높다 (총합의 정확도 암의 정확도는 A가 월등함)


#### precision(정밀도)
- 모델의 예측값이 얼마나 정확하게 예측됐는지를 나타낸다.
    - 정밀도 : TP / TP+FP
- 예제로 이해해보기 
    - A병원의 암환자 정밀도 : 9/9+30 = 0.23
    - B병원의 암환자 정밀도 : 1/1+20 = 0.04
    - A병원의 정밀도가 높다 

#### recall(재현율)
- 실제값 중에서 모델이 검출한 실제값의 비율을 나타낸다.
- 걸러내야하는 것들중에 제대로 검출된 것의 비율을 의미한다.
    - 정밀도 : TP/TP+FN
- 예제로 이해해보기
    - A병원의 암환자 재현율 : 9/9+1 = 0.9
    - B병원의 암환자 재현율 : 1/1+9 = 0.1
    - A병원의 재현율이 훨씬 높다 (암 환자의 경우 발견해내는것이 중요한 질병이기 때문에 무조건 A병원에 가야한다.)

### 재현율과 정밀도 언제 무엇을 써야할까?

- 두 가지는 사용하는 경우에 따라서 써야할게 다르다.
- 재현율이 중요한 경우 
    - 실제 positive 데이터를 negative로 잘못 판단하면 안되는경우 (암진단, 등등)
    - 병 진단 : 실제 양성인데 음성으로 판단하면 병을 키울 수 있다.
- 정밀도가 중요한 경우
    - 실제 negative 데이터를 positive로 잘못 판단하면 안되는경우 
    - 스팸메일 : 실제 스팸메일이 아닌데 스팸메일로 판단하면 메일을 받지 못할 수 있다.

- F-score
    - 정밀도와 재현율이 한쪽으로 치우치지 않았을때 높은 점수가 나온다.
    - precision과 recall을 이용해 계산가능
- 임계값
    - 모델을 분류에서 확률(0~1)또는 음수에서 양수사이에 실수를 예측값으로 출력
    - 사이킷런에선 predict_probar, decision_function으로 쓸 수 있다.
    - predict_probar: 0.5이상이면 1로 예측
    - decision_function : 0 이상이면 1로 예측 

## 모델 성능 평가지표 (Metric)
- 실제값과 모델에 의해 예측된 값을 비교하여 모델의 성능을 측정하는 것 
- 모델 평가 목적 
    - overfiting 방지 
- 분류 문제 성능 측정 
    - 실제값과 예측값이 일치하는 개수를 통해 측정 

### ROC
- FPR을 X축으로, TPR을 Y축으로 해서 둘간의 관계를 표현한 그래프
    - FPR = FP / (FP+TN)
    - TPR = TP / (FN+TP)
- FPR이 천천히 증가하면서 TPR이 빨리 증가하는 모델을 만들어야한다.
- 면적이 클수록 좋은 모델이다.

![ROC.png](https://miro.medium.com/max/640/1*pk05QGzoWhCgRiiFbz-oKQ.webp)

AUROC
- ROC Curved의 밑부분 면적
- 넓을수록 성능이 좋다 (1에 가까울수록..)
    - 0.5~0.7 (poor model)
    - 0.7~0.8 (fare model)
    - 0.8~0.9 (good model)
    - 0.9~1.0 (excellent model)

# 다중분류 Metric


### F1-score (다중분류)

- micro 
    - 전체 클래스에 대해 TP/FP/FN 구한뒤에 F1-score를 계산
- macro    
    - 각 클래스에 대해 F1-score 구한뒤 산술 평균(주어진 값을 개수로 나누기)
- weighted
    - 각 클래스에 대해 F1-score 구한뒤 클래스 차지하는 비율에 따라 가중평균 (가중치를 반영하여 구한 평균값)

### 소프트맥스 함수 (softmax Function)
- 각 클래스에 대한 확률을 출력
- 입력받은 값들을 0~1사이의 값들로 모두 정규화하면 출력값들의 합은 항상 1이 되는 특성을 가졌다.

### Cross_entropy
- 모델이 예측한 확률값을 반영해서 평가
- 0에 가까울수록 좋은 모델
- 정답에 해당하는 확률 값들을 음의 로그함수에 넣어서 나온값들의 평균으로 평가한다.

## 데이터 깔끔하게 전처리 해보기

- 데이터의 info()값을 보았을때 결측치가 존재할경우 어떻게 해결할까?

## 결측치

### 통계치를 이용해 결측값 채우기
- Pandas를 이용한 방법

```python
age_median = x["age"].median() # x데이터 프레임의 age컬럼의 중간값을 구한다.
x["age"].fillna(age_median, inplace=True) # age컬럼의 결측치를 중간값으로 채운다.
```


### Encoding 
- 범주형 인코딩 
1.  라벨 인코딩    
```python
items = ['사과', '딸기', '포도', '배', '망고', '망고', '바나나', '바나나']
ncoder = LabelEncoder() # 1. 객체 생성
encoder.fit(items) # 객체 fit encoding 대상에 대해 fit 
labels = encoder.transform(items)
print('Label tranform : ', labels)       
print('Encoding Class : ', encoder.classes_)
print('Inverse Transform : ', encoder.inverse_transform([0, 1, 4, 5, 3, 3, 2, 2]))
Label tranform :  [4 0 5 3 1 1 2 2]
Encoding Class :  ['딸기' '망고' '바나나' '배' '사과' '포도']
Inverse Transform :  ['딸기' '망고' '사과' '포도' '배' '배' '바나나' '바나나']
```


2. 원핫 인코딩
- 0,1로 만들어주는 데이터 (라벨인코더를 쓰고 만들어주기도 함 )

### Feature Scaling


- Standardization (표준화)
    - 평균과 표준편차를 이용하여 평균을 0, 분산을 1인 값으로 변환
                                       

- Min-Maxscaling 
    - Features의 최소값, 최대값을 이용해서 0을 정규화한다.

## 서포트 벡터 머신

- 회귀 분류 이상치 탐지에 쓰인다. 
- 마진이란?
    - 각 class에서 가장 가까운 관측치 사이의 거리를 나타낸다.
    - 마진은 기울기로 표현 가능하다.
- 하드마진과 소프트마진
    - 하드마진 방식은 매우 엄격하게 두 개의 클래스를 분리하는 분리 초평면을 구하는 방법이다.
    - 모든 트레이닝 셋은 분리 초평면을 두고 한 클래스에 속해야 한다.   
    - 소프트마진은 경계선에 약간의 차이를 벌리는걸 하드마진과의 차이라고 볼 수 있다. 


### SVM을 사용한 분류 모델 (SVC)
```python
model = SVC()
model.fit(x_train, y_train)
# SVR 커널 종류는 ‘linear’, ‘poly’, ‘rbf’, ‘sigmoid’, ‘precomputed’이 있다. 
```


### 효율적인 SVM 사용방법 
- svm은 데이터가 정규화 되어야 좋은 성능을 보임
- 주로 특성값을 (0,1)의 사이로 맞추는 방법을 사용
- 스텐다드 스케일러와 민맥스 스케일러를 사용해 정규화

## KNN

